src/Compilers/CSharp/Test/Semantic/Semantics/RefEscapingTests.cs

```diff
@@ -0,0 +1,1202 @@
+                //             var dummy2 = new Program(ref rOuter, ref rInner);
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "new Program(ref rOuter, ref rInner)").WithArguments("Program.Program(ref Program.S1, ref Program.S1)", "arg2").WithLocation(19, 26),
+                // (22,42): error CS8168: Cannot return local 'inner' by reference because it is not a ref local
+                //             var dummy3 = new Program(ref inner, ref rOuter);
+                Diagnostic(ErrorCode.ERR_RefReturnLocal, "inner").WithArguments("inner").WithLocation(22, 42),
+                // (22,26): error CS8524: This combination of arguments to 'Program.Program(ref int, ref Program.S1)' is disallowed because it may expose variables referenced by parameter 'arg1' outside of their declaration scope
+                //             var dummy3 = new Program(ref inner, ref rOuter);
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "new Program(ref inner, ref rOuter)").WithArguments("Program.Program(ref int, ref Program.S1)", "arg1").WithLocation(22, 26)
+            );
+        }
+
+        [Fact()]
+        public void RefLikeEscapeMixingCallOptionalIn()
+        {
+            var text = @"
+    class Program
+    {
+        static void Main()
+        {
+        }
+
+        void Test1()
+        {
+            S1 rOuter = default;
+
+            int inner = 1;
+            S1 rInner = MayWrap(inner);
+
+            // error
+            MayAssign(ref rOuter);
+
+            // valid, optional arg is of the same escape level
+            MayAssign(ref rInner);
+        }
+
+        static void MayAssign(ref S1 arg1, in int arg2 = 42)
+        {
+            arg1 = MayWrap(arg2);
+        }
+
+        static S1 MayWrap(in int arg)
+        {
+            return default;
+        }
+
+        ref struct S1
+        {
+        }
+    }
+";
+            CreateStandardCompilation(text).VerifyDiagnostics(
+                // (16,13): error CS8524: This combination of arguments to 'Program.MayAssign(ref Program.S1, in int)' is disallowed because it may expose variables referenced by parameter 'arg2' outside of their declaration scope
+                //             MayAssign(ref rOuter);
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "MayAssign(ref rOuter)").WithArguments("Program.MayAssign(ref Program.S1, in int)", "arg2").WithLocation(16, 13)
+            );
+        }
+
+        [Fact()]
+        public void MismatchedRefTernaryEscape()
+        {
+            var text = @"
+class Program
+{
+    static void Main()
+    {
+    }
+
+    public static int field = 1;
+
+    void Test1()
+    {
+        var local = 42;
+        ref var ternary1 = ref true ? ref field : ref local;
+
+        ref var lr = ref local;
```

Comment:
Did you mean to use `lr` and `fr`?
Should we have un-used warnings for those?

Reply:
Yes, that was the idea.

---
In reply to: [135400610](https://github.com/dotnet/roslyn/pull/21334#discussion_r135400610) [](ancestors = 135400610)